sycl/include/sycl/ext/oneapi/reduction.hpp

```diff
@@ -504,78 +496,95 @@ template <typename T, class BinaryOperation> class reduction_impl_common {
   bool InitializeToIdentity;
 };
 
-/// Types representing specific reduction algorithms
-/// Enables reduction_impl_algo to take additional algorithm-specific templates
-template <bool IsUSM, access::placeholder IsPlaceholder, int AccessorDims>
-class default_reduction_algorithm {};
+template <class T> struct is_rw_acc_t : public std::false_type {};
 
-/// Templated class for implementations of specific reduction algorithms
-template <typename T, class BinaryOperation, int Dims, size_t Extent,
-          class Algorithm>
-class reduction_impl_algo;
+template <class T, int AccessorDims, access::placeholder IsPlaceholder,
+          typename PropList>
+struct is_rw_acc_t<accessor<T, AccessorDims, access::mode::read_write,
+                            access::target::device, IsPlaceholder, PropList>>
+    : public std::true_type {};
+
+template <class T> struct is_dw_acc_t : public std::false_type {};
+
+template <class T, int AccessorDims, access::placeholder IsPlaceholder,
+          typename PropList>
+struct is_dw_acc_t<accessor<T, AccessorDims, access::mode::discard_write,
+                            access::target::device, IsPlaceholder, PropList>>
+    : public std::true_type {};
+
+template <class T> struct is_placeholder_t : public std::false_type {};
+
+template <class T, int AccessorDims, access::mode Mode, typename PropList>
+struct is_placeholder_t<accessor<T, AccessorDims, Mode, access::target::device,
+                                 access::placeholder::true_t, PropList>>
+    : public std::true_type {};
+
+template <class T> struct accessor_dim_t {
+  static constexpr int value = 1;
+};
```

Comment:
USM goes through this route.

Reply:
Ah! Maybe we should avoid the word `accessor` in this then?

Reply:
Ideally, yes, but it's not something I invented/regressed here. Just re-phrasing what was written before me. I hope a better way would be to fully decouple internal accessors from the user's reduction variable, but we're not there yet. 

IIUC, the reason we have it like this is to implement the single work group optimization where we use output variable as partial sums in the kernel and need the types matching. I believe passing both accessor *and* variable and doing it via `if` inside the kernel is almost as performant, but that would be a different patch. I also have one draft that does that in one of the kernels.

That said, I'd prefer to leave that for a future patch.

Reply:
Also, it's not about *incoming* accessor, it's about the one we will be using in the kernel, which matches the incoming one in case that's an accessor.

Reply:
> Also, it's not about _incoming_ accessor, it's about the one we will be using in the kernel, which matches the incoming one in case that's an accessor.

Personally I prefer having the name of traits like these say what they do rather than what they are being used for elsewhere. Makes it easier to read if you fall over it randomly and it makes it easier to use for other purposes later.

That said, if you believe it will be removed/renamed in a later patch, I am okay with the current version of it.

Reply:
> I prefer having the name of traits

Ah, now I understand your concern. What if I'll change the trait's name to `compute_accessor_dim_t`? Would that address it?

Reply:
I was thinking something like `data_dim_t` because it gets the dimensions of the accessible data, either from an accessor (N dimensional) or a pointer (1-dimensional), but I can live with `compute_accessor_dim_t`. 😄